README
A port of mobility-fw2.0p3 for OMNeT++ 4.0 simulator
====================================================
by Jerome Rousselot <jerome.rousselot@csem.ch> and Rudolf Hornig <rudi@omnetpp.org>

This is an unofficial (not supported by its original authors) port of the 
Mobility Framework 2.0p3 to OMNeT++ 4.0.

Its latest version can be found at github.com:
http://github.com/mobility-fw/mf-opp4/tree/master

and some information is available at the wiki page:
http://wiki.github.com/mobility-fw/mf-opp4

Two branches are available: master and ext.

master is a "pure" port of mf2.

ext includes additional modules from other authors. Contributions and bug fixes are welcome.
An example network was added in networks/ieee802154 to demonstrate the use of the
high precision multiple access interference PHY SnrEvalRadioAccNoise3, the CSMA protocol
as defined in IEEE 802.15.4 (csma.ned) and the WiseRoute routing layer (wiseroute.ned).

Several modules in the ext branch have been partially financed by the European Commission 
under the Framework 6 IST Project "Wirelessly Accessible Sensor Populations" (WASP) 
under contract IST-034963.

To run examples in the "networks" folder. Use the "run" script in the example dirs
Note that the "networks", folder does not contain any .cc files so it is not added to the makefiles.

Changes
-------
In contrib, the P2PLayer does not use the sendDirect method anymore.
The core/include and contrib/include directories have been removed, as well as
the core/lib and contrib/lib directories, to simplify the build process.
The libraries are now in core/libmfcore.so and in contrib/libmfcontrib.so.

Contents
========
 - Purpose
 - Requirements
 - Copyright
 - Installation
 - Getting started
 - Release notes
 - Legal stuff


Purpose
=======
The mobility framework provides abstractions needed to develop network
protocols for non-stationary wireless nodes. 


Requirements
============
This version of the Mobility Framework requires OMNeT++ 4.0 or higher.
The only platform it has been tested on is Linux (2.6.xx kernel, g++ 4.3.x). 


COPYRIGHT
=========
Copyright (C) 2004  Telecommunication Networks Group at Technische
Universitaet Berlin, Germany

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


Installation 
============
1. Unpack the archive (you have probably already done so)
   tar xzf mobility-fw<VERSION>.tgz

2. run the mkmk shell script
   ./mkmk

3. make all 


Importing the project into the OMNeT++ 4 IDE
============================================
After starting the OMNeT++ IDE, you can import the project by right-clicking
in the project list and choosing the "Import..." wizard, and 
"General->Existing project into Workspace".  Select the directory containing 
the branch as root directory. You should be able to build it by pressing CTRL-B. 


Getting started
===============
 
This will "install" the library in the lib directories and the
include files in the include directories of the MF.

To see if it runs, try the networks in the networks directory.
Select or open an inifile and press the Run button in the IDE.

Copy the template from the template directory to a directory of
your choice -- it has its own INSTALL instructions to get you
started quickly with your own simulation. NOTE: The template 
directory is not fully adapted to MF 2.0. So not all the stuff
will work!


Release notes
=============
* 2.0p3 for OMNeT++ 4
The same version for OMNeT++ 4 converted by Jerome Rousselot 
<jerome.rousselot@csem.ch> and Rudolf Hornig <rudi@omnetpp.org>
Changes: In contrib, the P2PLayer does not use the sendDirect method anymore.
The core/include and contrib/include directories have been removed, as well as
the core/lib and contrib/lib directories, to simplify the build process.
The libraries are now in core/libmfcore.so and in contrib/libmfcontrib.so.

* 2.0p3
just another preview. Basically just contains several 802.11 bug fixes.
Re-added old manual of MF v1

* 2.0p2:
Preview no.2. Main enhancements (apart from some minor bug fixes) are:
- windows make system should work know. See the install directions for 
  details
- MAC OS make system added. MF can be installed on MAC OS either with or 
  without shared libraries. NOT TESTED!
- 802.11 should work again - at least with the basic bitrate. However,
  the automatic bitrate adaption still crashes.
- We started to adapt the template folder to MF 2.0 - not finished yet!

* 2.0p1:
Preview of a MF with major restructuring and redesign. As with previews there
may still be bugs and not everything has been thorouly tested yet. Especially 
the 802.11 network still CRASHES! The manual has not been updated yet, please 
use the doxygen documentation for detailed information.
Key changes include:
- enhanced efficiency of the code (speed and memory consumption), up to 
  thousands of nodes can now be simulated on standard PCs
  - added sendDirect option (less gates -> less memory)
  - added grid to channel control to speed up connection management
- more user implementation flexibility
  - reduced core MF (only contains basic mobility, channel control and modules
    up to the MAC)
  - cleaner possibility to add / modify control infos to messages
  - moved radio related stuff into a seperate radio module
  - added arp functionality
- completely redesigned blackboard
  - only publisher can change published items
  - items are identified by their class instead of strings (speed!)
  - items can be subclassed and extended
- better mobility support
  - complete border handling implemented
  - supported policies are: reflect, wrap, place randomly and raise error
- bug fixes
  - several bugs removed from aloha, csma and snr-eval modules
  - started to debug 802.11 - NOT FINISHED yet!


Legal Stuff
===========
The copying, modifications, redistribution and so on is covered by the
GNU General Public License, see the file COPYING for details.

The Telecommunication Networks Group at the Technische Universitaet
Berlin does not provide any warranty for the Mobility Framework -- 
you did not pay for it! Further details can be found in the file 
COPYING.

As a good practice, you should acknowledge the work done at the
Telecommunication Networks Group at the Technische Universitaet
Berlin, just as you do if you cite a scientific work.


ChangeLog
2008-12-20: Rudolf Hornig modified the build system to match the recommended 
            build structure for OMNeT++ 4

2008-12-18: Jerome Rousselot ported to OMNeT++ 4.0

2007-01-12  ------ mobility-fw2.0p3 released ------

2007-01-12:	Daniel Willkomm
	* Version: update to 2.0p3
	* README: added release notes for 2.0p3

2006-08-11  ------ mobility-fw2.0p2 released ------

2006-08-11:	Daniel Willkomm
	* Version: update to 2.0p2
	* README: added release notes for 2.0p2 and updated install instructions

2006-07-28:	Daniel Willkomm
	* README: updated windows installation and added MAC OS installation
	instructions
	* makemakefiles: removed snrTest network
	* mkmk: copy linux lib makefrags before makeing makfiles

2006-07-27:	Daniel Willkomm
	* makemakefiles: linux and windows Makefile to create all makefiles for 
	the MF -> no need to maintain two seperate files for windows and linux
	* Makefile.linux Makefile.mac Makefile.win: OP specific Makefile setings to
	create all makefiles for the MF
	* mfconfig.vc: removed
	* mkmk mkmk.cmd: reduced to calling of make command with the appropriate
	Makefile.OP to create all makefiles for the MF
	* mkmk.mac: added to create makefiles for MAC OS without shared libraries

2006-07-21:	Daniel Willkomm
	* mkmk.cmd: adapted windows make system to MF2.0
	* mfconfig.vc: adapted windows make system to MF2.0
	* makefrag.vc: added quotes to echo command

2006-07-06  ------ mobility-fw2.0p1 released ------

2006-07-06:	Daniel Willkomm
	* mkmk.cmd: adapted windows script to new MF - not tested!
	* makefrag.vc: adapted to new MF - not tested!
	* makefrag: added doxy and neddoc targets
	* README: added release notes

2006-07-05:	Daniel Willkomm
	* mkmk: changed testSuite Makefiles to cmd mode, removed coreTest target

2006-07-04:	Daniel Willkomm
	* AUTHORS: updated author information
	* README: updated install information

2006-06-30:	Daniel Willkomm
	* mkmk: added aloha network target

2006-06-29:	Daniel Willkomm
	* mkmk: removed development target, added top level networks target, added
	snrTest and baseStationNet target

2006-06-12:	Daniel Willkomm
	* mkmk: added development/mobility to script

2005-12-01 Daniel Willkomm
	* mkmk: added devel target for the lib dir, changed network targets to also
	include the devel lib, added development/netwLayer target
	* makefrag: added devel_lib target

2005-11-23: Daniel Willkomm
	* mkmk: added contrib/utils target

2005-10-19: Daniel Willkomm
	* Makefile: removed; is created by mkmk
	* makefrag: added for top level Makefile
	* mkmk: updated to create top level Makefile and omnetppconfig

2005-10-18: Daniel Willkomm
	* mkmk: started to update to new structure - not finished yet!!!
	* Makefile: changed to work with restructuring, core and contrib can be
	successfully compiled again

2005-01-18  ------ mobility-fw1_0a5 released ------

2004-11-22  ------ mobility-fw1.0a4 released ------

2004-11-05  ------ mobility-fw1.0a4 released ------

2004-10-20  ------ mobility-fw1.0a3 released ------

2004-10-12	Daniel Willkomm
	* mkmk: added script to create all Makefiles
	* mkmk.cmd: added -b and -x option to script. cleaning up.
	* README: removed windows symlink stuff. Added mkmk.cmd advise for
	windows users in case the MF does not compile
	* Makefile: recreated all Makefiles

2004-10-04	Daniel Willkomm
	* Makefile: deleted the symlink stuff for the networks dirs, as it
	is not needed anymore (since omnet++3.0a7).

2004-09-13	Daniel Willkomm
	* makefrag.vc, Makefile.vc: added line to copy BasicModule.h into
	the include dir.
	* README: updated windows install directions. nmake needs the '/f
	Makefile.vc' option otherwise it uses the linux Makefile.

2004-08-23	Daniel Willkomm
	* mkmk.cmd: corrected the lib Makefile to also include the 'core'
	dir.

2004-08-18  ------ mobility-fw1.0a2 released ------

2004-08-17	Daniel Willkomm
	* README: added windows install directions
	* Makefiles.vc: added windows Makefiles (Makefile.vc) to each
	directory. Windows Makefiles were provided by Andras (thanks!)
	* makefrag.vc: contains make stuff from linux Makefile to be
	included into the top-level windows MAkefile.vc
	* mkmk.cmd: script to recursively rebuild all Makefile.vc's

2004-08-16	Daniel Willkomm
	* Makefile: adjusted the Makefile to be somewhat similar to the
	windows Makefile.vc. Removed obsolete targets. Simplified doxy
	generation

	* Makefiles: rebuild all Makefiles

2004-07-23  ------ mobility-fw1.0a1 released ------


contrib/nic/macLayer/csma.h
/* -*- mode:c++ -*- ********************************************************
 * file:        csma.h
 *
  * author:     Jerome Rousselot, Marcel Steine, Amre El-Hoiydi,
 *				Marc Loebbers, Yosia Hadisusanto
 *
 * copyright:	(C) 2007-2009 CSEM SA
 * 				(C) 2009 T.U. Eindhoven
 *				(C) 2004 Telecommunication Networks Group (TKN) at
 *              Technische Universitaet Berlin, Germany.
 *
 *              This program is free software; you can redistribute it
 *              and/or modify it under the terms of the GNU General Public
 *              License as published by the Free Software Foundation; either
 *              version 2 of the License, or (at your option) any later
 *              version.
 *              For further information see file COPYING
 *              in the top level directory
 *
 * Funding: This work was partially financed by the European Commission under the
 * Framework 6 IST Project "Wirelessly Accessible Sensor Populations"
 * (WASP) under contract IST-034963.
 ***************************************************************************
 * part of:    Modifications to the MF-2 framework by CSEM
 **************************************************************************/


#ifndef csma_H
#define csma_H


#include <string>
#include <sstream>
#include <vector>
#include <list>
#include <ActiveChannel.h>
#include <RadioAccNoise3State.h>
#include <RSSI.h>
#include "BaseMacLayer.h"
#include <Bitrate.h>
#include <DroppedPacket.h>
#include <BasicMobility.h>
#include <Blackboard.h>
#include <MacPkt_m.h>
#include "MacControlInfo.h"
#include <SingleChannelRadioAccNoise3.h>
#include "RadioAccNoise3PhyControlInfo.h"
#include "ConstsAccNoise3.h"

/**
 * @class csma
 * @ingroup macLayer
 * @author Jerome Rousselot, Amre El-Hoiydi, Marc Loebbers, Yosia Hadisusanto, Andreas Koepke
 */
class  csma : public BaseMacLayer
{
  public:


	~csma();

    /** @brief Initialization of the module and some variables*/
    virtual void initialize(int);

    /** @brief Delete all dynamically allocated objects of the module*/
    virtual void finish();

    /** @brief Handle messages from lower layer */
    virtual void handleLowerMsg(cMessage*);

    /** @brief Handle messages from upper layer */
    virtual void handleUpperMsg(cMessage*);

    /** @brief Handle self messages such as timers */
    virtual void handleSelfMsg(cMessage*);

    /** @brief Handle control messages from lower layer */
    virtual void handleLowerControl(cMessage *msg);

    /** @brief Called by the Blackboard whenever a change occurs we're interested in */
    virtual void receiveBBItem(int category, const BBItem *details, int scopeModuleId);

  protected:
    typedef std::list<MacPkt*> MacQueue;

  long nbTxFrames;
  long nbRxFrames;
  long nbMissedAcks;
  long nbRecvdAcks;
  long nbDroppedFrames;
  long nbTxAcks;
  long nbDuplicates;
  long nbBackoffs;
  double backoffValues;
  bool stats;
  bool trace;

    /** @brief MAC states
     * see states diagram.
     */

    enum t_mac_states {
        IDLE_1=1,
        BACKOFF_2,
        CCA_3,
        TRANSMITFRAME_4,
        WAITACK_5,
        WAITSIFS_6,
        TRANSMITACK_7

    };

    /*************************************************************/
    /****************** TYPES ************************************/
    /*************************************************************/

    enum t_mac_timer {
      TIMER_NULL=0,
      TIMER_BACKOFF,
      TIMER_CCA,
      TIMER_SIFS,
      TIMER_RX_ACK,
    };

    cMessage * backoffTimer, * ccaTimer, * txTimer, * sifsTimer, * rxAckTimer;

    enum t_mac_event {
      EV_SEND_REQUEST=1,                   // 1, 11, 20, 21, 22
      EV_TIMER_BACKOFF,                    // 2, 7, 14, 15
      EV_FRAME_TRANSMITTED,                // 4, 19
      EV_ACK_RECEIVED,                     // 5
      EV_ACK_TIMEOUT,                      // 12
      EV_FRAME_RECEIVED,                   // 15, 26
      EV_DUPLICATE_RECEIVED,
      EV_TIMER_SIFS,                       // 17
      EV_BROADCAST_RECEIVED, 		   // 23, 24
      EV_TIMER_CCA
    };

    enum t_csma_frame_types {
    	DATA,
    	ACK
    };
    enum t_mac_carrier_sensed {
      CHANNEL_BUSY=1,
      CHANNEL_FREE
    } ;

    enum t_mac_status {
      STATUS_OK=1,
      STATUS_ERROR,
      STATUS_RX_ERROR,
      STATUS_RX_TIMEOUT,
      STATUS_FRAME_TO_PROCESS,
      STATUS_NO_FRAME_TO_PROCESS,
      STATUS_FRAME_TRANSMITTED
    };

    /** @brief keep track of MAC state */
    t_mac_states macState;
    t_mac_status status;

    /** @brief Current state of active channel (radio), set using radio, updated via BB */
    RadioAccNoise3State::States radioState;
    /** @brief category number given by bb for RadioState */
    int catRadioState;

    /** @brief Last RSSI level */
    double rssi;
    /** @brief category number given by bb for RSSI */
    int catRSSI;

    /** @brief RSSI level where medium is considered busy */
    double busyRSSI;



    /** @brief Maximum time between a packet and its ACK
     *
     * Usually this is slightly more then the tx-rx turnaround time
     * The channel should stay clear within this period of time.
     */
    simtime_t sifs;
    simtime_t macAckWaitDuration;
    bool transmissionAttemptInterruptedByRx;
    /** @brief CCA detection time */
    simtime_t ccaDetectionTime;
    /** @brief Time to setup radio from sleep to Rx state */
    simtime_t rxSetupTime;
    /** @brief Time to switch radio from Rx to Tx state */
    simtime_t aTurnaroundTime;
    /** @ brief minimum IEEE 802.15.4-2006 backoff exponent */
    int macMinBE;
    /** @brief maximum IEEE 802.15.4-2006 backoff exponent */
    int macMaxBE;
    /** @brief maximum number of backoffs before frame drop */
    int macMaxCSMABackoffs;
    /** @brief maximum number of frame retransmissions without ack */
    unsigned int macMaxFrameRetries;
    /** @brief base time unit for calculating backoff durations */
    simtime_t aUnitBackoffPeriod;


    /** @brief cached pointer to radio module */
    SingleChannelRadioAccNoise3* radio;

    /** @brief number of backoff performed until now for current frame */
    int NB;
    /** @brief current value of backoff exponent for current frame */
    int BE;
    /** @brief cached pointer to mobility module */
    BasicMobility * mobility;
    /** @brief A queue to store packets from upper layer in case another
    packet is still waiting for transmission..*/
    MacQueue macQueue;

    /** @brief length of the queue*/
    unsigned int queueLength;

    /** @brief count the number of tx attempts
     *
     * This holds the number of transmission attempts for the current frame.
     */
    unsigned int txAttempts;


    /** @brief the bit rate at which we transmit */
    double bitrate;

    double backoffWindow;

    /** @brief initial contention window size */
    double initialCW;

    /** @brief Inspect reasons for dropped packets */
    DroppedPacket droppedPacket;

    /** @brief plus category from BB */
    int catDroppedPacket;

    /** @brief publish dropped packets nic wide */
    int nicId;

    int headerLength;

    int ackLength;


private:
	// FSM functions
	void fsmError(t_mac_event event, cMessage *msg);
	void executeMac(t_mac_event event, cMessage *msg);
	void updateStatusIdle(t_mac_event event, cMessage *msg);
	void updateStatusBackoff(t_mac_event event, cMessage *msg);
	void updateStatusCCA(t_mac_event event, cMessage *msg);
	void updateStatusTransmitFrame(t_mac_event event, cMessage *msg);
	void updateStatusWaitAck(t_mac_event event, cMessage *msg);
	void updateStatusSIFS(t_mac_event event, cMessage *msg);
	void updateStatusTransmitAck(t_mac_event event, cMessage *msg);
	void updateStatusNotIdle(cMessage *msg);
	void manageQueue();
	void updateMacState(t_mac_states newMacState);
	void manageMissingAck(t_mac_event event, cMessage *msg);
	void startTimer(t_mac_timer timer);
	double scheduleBackoff();
	cMessage *decapsMsg(MacPkt * macPkt);
	MacPkt * ackMessage;

	//sequence number for sending, map for the general case with more senders
	//also in initialisation phase multiple potential parents
	map<int, unsigned long> SeqNrParent; //parent -> sequence number

	//sequence numbers for receiving
	map<int, unsigned long> SeqNrChild; //child -> sequence number

};

#endif

